### PR TITLE
Second commit of speculative execution in MR mode

### PR DESCRIPTION
**Pull request by ydai1124**
Design doc: https://docs.google.com/document/d/1Vf5A7pHbFKwr7-Bbw2X3zBBVZSvT7HXEPE6bsi9qGco/edit#

What this PR contains:
1. Modified `Task` to contain the real `run` and `commit` method.
2. Modified `AbstractJobLauncher` to create `GobblinMultiTaskAttempt`, and use the `run` and `commit` steps in `GobblinMultiTaskAttempt`.
3. Modified `MRJobLauncher` to accommodate the situation with speculative execution enabled.
4. Modified `TaskStateTracker` to have two callback functions: `onTaskRunCompletion` and`onTaskCommitCompletion`.

### CODE REVIEW

**Comment by ydai1124** [comment]
@chavdar @ibuenros  can you take a look? Thanks!

**Comment by chavdar** [comment]
There are FindBugs failures:

> Execution failed for task ':gobblin-core:findbugsMain'

**Comment by ydai1124** [comment]
@chavdar Couldn't reproduce locally. Re-launched the travis build and it succeeded.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractor.java [hunk 1]
```diff
@@ -62,10 +62,8 @@ public FileAwareInputStream readRecord(@Deprecated FileAwareInputStream reuse)
       throws DataRecordException, IOException {
 
     if (!this.recordRead) {
-      Configuration conf = HadoopUtils.newConfiguration();
-      FileSystem fsFromFile = this.file.getOrigin().getPath().getFileSystem(conf);
       this.recordRead = true;
-      return new FileAwareInputStream(this.file, fsFromFile.open(this.file.getFileStatus().getPath()));
+      return new FileAwareInputStream(this.file, this.fs.open(this.file.getFileStatus().getPath()));
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractor.java]: This will break pull mode from multiple `FileSystem`s. Why is this change needed?

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java [hunk 2]
```diff
@@ -575,7 +529,69 @@ private void addConstructsFinalStateToTaskState(InstrumentedExtractorBase<?, ?>
    * 3. Check whether to publish data in task.
    */
   public void commit() {
-    // TODO- refactor the current Task.run() method to run & commit phases, and add commit content here.
+    try {
+      // Check if all forks succeeded
+      boolean allForksSucceeded = true;
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: shouldn't the success check happen in the `run()` method?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractor.java (hunk 1 above)
Comment (ydai1124) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractor.java]: This change was introduced in #1306 and broke lumos-ng puller. The source filesystem cannot be got from an empty `configuration`. 

I was unaware of the multiple `FileSystem` mode. One possible solution may be instead of using `path.getFs`, will `copyfile.getOriginFs` work for the multiple `FileSystem` mode? Another solution could be creating one more constructor for `FileAwareInputStreamExtractor`, which can pass the job `properties`/`state` in. Then `path.getFs` can have a `FileSystem` instance for `http` uri.

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java (hunk 2 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: This check is checking if `fork.commit()` succeeds. The main job of `fork.commit()` is to run `writer.commit()` method to move data from staging to writers' final output. 

The `commit` phase, especially in speculative execution mode, should be something executed only once. So I moved `fork.commit` to the `commit` phase.

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java [hunk 3]
```diff
@@ -117,13 +121,22 @@ public Task(TaskContext context, TaskStateTracker taskStateTracker, TaskExecutor
         closer.register(new InstrumentedExtractorDecorator<>(this.taskState, this.taskContext.getExtractor()));
 
     this.converter = closer.register(new MultiConverter(this.taskContext.getConverters()));
-
+    try {
+      this.rowChecker = closer.register(this.taskContext.getRowLevelPolicyChecker());
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: Why is this necessary? Shouldn't the converter instantiation be also in the try-catch block?

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java [hunk 4]
```diff
@@ -575,7 +529,69 @@ private void addConstructsFinalStateToTaskState(InstrumentedExtractorBase<?, ?>
...
+        }
+      }
+
+      if (allForksSucceeded) {
+        // Set the task state to SUCCESSFUL. The state is not set to COMMITTED
+        // as the data publisher will do that upon successful data publishing.
+        this.taskState.setWorkingState(WorkUnitState.WorkingState.SUCCESSFUL);
+      } else {
+        LOG.error(String.format("Not all forks of task %s succeeded", this.taskId));
+        this.taskState.setWorkingState(WorkUnitState.WorkingState.FAILED);
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: please use failTask() with a exception specifying which forks failed. This will ensure consistent behavior about what failure state is maintained in the task.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 5]
```diff
@@ -623,61 +625,42 @@ private EventSubmitter buildEventSubmitter(List<? extends Tag<?>> tags) {
    * @param taskExecutor a {@link TaskExecutor} for task execution
    * @param taskStateStore a {@link StateStore} for storing {@link TaskState}s
    * @param logger a {@link Logger} for logging
+   * @param shouldCommitTasks whether should commit {@link GobblinMultiTaskAttempt}
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: I think this parameter name and javadoc are misleading. Ultimately, all successful tasks should commit. The questions is whether they should commit immediately after each run or at the end. Please think about improving the naming. Using a enum instead of boolean can also improve readability.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java [hunk 6]
```diff
@@ -577,17 +581,53 @@ protected void setup(Context context) {
...
+              .put(context.getTaskAttemptID().toString(), gobblinMultiTaskAttempt);
+        }
       } finally {
-        this.cleanup(context);
+        CommitStep cleanUpCommitStep = new CommitStep() {
+
+          @Override
+          public boolean isCompleted()
+              throws IOException {
+            throw new UnsupportedOperationException("Not supported.");
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java]: What is the expected behavior here? Isn't it better to return "false" instead of throwing an exception?

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java [hunk 7]
```diff
@@ -577,17 +581,53 @@ protected void setup(Context context) {
...
+          public void execute()
+              throws IOException {
+            LOG.info("Starting the clean-up steps.");
+            try {
+              serviceManager.stopAsync().awaitStopped(5, TimeUnit.SECONDS);
+            } catch (TimeoutException te) {
+              // Ignored
+            } finally {
+              if (jobMetrics.isPresent()) {
+                // TODO: Add "committed" tag to final metrics.
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java]: Is this going to be fixed?

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java (hunk 6 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java]: It is used by "exactly once" semantics to check if a `commitstep` from a previous execution has been executed or not. Change to `return !serviceManager.isHealthy();` to return if `serviceManager` has stopped.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java (hunk 7 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java]: Yes. Next and last PR will be adding this tag, which is currently not supported by Gobblin metrics.

File: gobblin-runtime/src/main/java/gobblin/runtime/Task.java (hunk 3 above)
Comment (ydai1124) [inline gobblin-runtime/src/main/java/gobblin/runtime/Task.java]: Because only `getRowLevelPolicyChecker` throws exception in its method signature.

**Comment by ydai1124** [comment]
@chavdar @ibuenros comments addressed.
